### PR TITLE
[QEff Finetune]: Added support for gradient checkpointing in the finetuning script.

### PR DESCRIPTION
**Pull request by quic-meetkuma**
Added --gradient_checkpointing new CLI flag to enable this feature.
Currently this is enabled for all the HF models which has "supports_gradient_checkpointing" attribute set to True.

### CODE REVIEW

File: QEfficient/finetune/dataset/samsum_dataset.py [hunk 1]
```diff
@@ -46,3 +48,22 @@ def tokenize_add_label(sample):
...
+    attn_mask = pad_sequence(
+        [torch.tensor(b["attention_mask"], dtype=torch.int32) for b in batch], batch_first=True, padding_value=0
+    )
+    labels = pad_sequence(
+        [torch.tensor(b["labels"], dtype=torch.long) for b in batch], batch_first=True, padding_value=eos_token
+    )
+    return {"input_ids": input_ids, "attention_mask": attn_mask, "labels": labels}
+
+
+def get_samsum_collate_fn(dataset_processer, dataset_config):
```
Comment (quic-mamta) [inline QEfficient/finetune/dataset/samsum_dataset.py:68]: We can use this collate function for other datasets also.

File: QEfficient/finetune/utils/train_utils.py [hunk 2]
```diff
@@ -177,10 +177,7 @@ def train(
                         # adjust atol & rtol this as required
                         atol=1e-1,
                         use_ref_output_on_mismatch=True,
-                        # report all mismatches
-                        max_failures=None,
-                        # generate unittest for each op once
-                        repeat_same_op=True,
+                        filter_config=qaic_debug.DispatchFilterConfig.default(device, opwise_limit=10),
```
Comment (quic-mamta) [inline QEfficient/finetune/utils/train_utils.py:180]: can you please explain this change?

File: QEfficient/finetune/dataset/samsum_dataset.py (hunk 1 above)
Comment (quic-meetkuma) [reply QEfficient/finetune/dataset/samsum_dataset.py:68]: Yes, good suggestion. We can take this as a part of refactoring PRs.

File: QEfficient/finetune/utils/train_utils.py (hunk 2 above)
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py:180]: This is based on recent changes to API. I will try to find a reference code from eager team.

**Review by quic-meetkuma — COMMENTED** [review]
I will add eager team's code reference for qaic_debug API changes.

File: QEfficient/finetune/utils/train_utils.py [hunk 3]
```diff
@@ -282,6 +280,12 @@ def train(
             else:
                 train_epoch_loss = total_loss / len(train_dataloader)
 
+        if train_config.enable_ddp:
+            # Get the correct train loss from all the nodes.
+            dist.barrier()
+            dist.all_reduce(train_epoch_loss, op=dist.ReduceOp.SUM)
+            train_epoch_loss /= dist.get_world_size()
+
```
Comment (quic-swatia) [inline QEfficient/finetune/utils/train_utils.py:288]: This will save the same recomputed  value of train_epoch_loss and subsequently train_perplexity in all N metrics_data_*.json file for NxDDP. It is not a good idea to lose those individual values as it will also match with the  train_step_loss values as per calculation.

**Review by quic-swatia — CHANGES_REQUESTED** [review]

File: QEfficient/finetune/utils/train_utils.py (hunk 3 above)
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py:288]: That is correct. I will revert it.

File: QEfficient/finetune/utils/train_utils.py [hunk 4]
```diff
@@ -298,6 +302,7 @@ def train(
                 )
                 dist.barrier()
                 dist.all_reduce(eval_epoch_loss, op=dist.ReduceOp.SUM)
+                eval_epoch_loss /= dist.get_world_size()
```
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py:305]: Here, the eval_epoch_loss is already been reduce-summed. I think we need to get rid of line no 303 and 304 to have different values in each json file. I will any ways remove L305.

**Review by quic-meetkuma — COMMENTED** [review]
Need to fix couple of comments.

File: QEfficient/finetune/utils/train_utils.py (hunk 2 above)
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py]: Confirmed with eager team that this is not required. Reverted the added line in latest commit.

File: QEfficient/finetune/utils/train_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (quic-mamta) [reply QEfficient/finetune/utils/train_utils.py]: yes we should remove line 303 and 304 as well.

[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py]: Removed in the latest commit.

[the remark this answers is not in the document]
Comment (quic-swatia) [reply QEfficient/finetune/utils/train_utils.py]: dist.barrier() is still there with the latest commit.

File: QEfficient/finetune/utils/train_utils.py [hunk 5]
```diff
@@ -177,10 +177,6 @@ def train(
                         # adjust atol & rtol this as required
                         atol=1e-1,
                         use_ref_output_on_mismatch=True,
-                        # report all mismatches
-                        max_failures=None,
-                        # generate unittest for each op once
```
Comment (quic-swatia) [inline QEfficient/finetune/utils/train_utils.py:182]: Please check once with pytorch eager team if they want these lines to be removed in general.

File: QEfficient/cloud/finetune.py [hunk 6]
```diff
@@ -103,6 +103,18 @@ def main(**kwargs):
...
+    # Note: Need to call this before calling PeftModel.from_pretrained or get_peft_model.
+    # Because, both makes model.is_gradient_checkpointing = True which is used in peft library to
+    # apply gradient checkpointing related hooks to the input embeddings. Without this we will get
+    # "No inf checks were recorded for this optimizer." error.
+    # Enable gradient checkpointing
+    if train_config.gradient_checkpointing:
+        # Note: below attribute and method is only available in HuggingFace Transformer models.
+        if model.supports_gradient_checkpointing:
+            model.gradient_checkpointing_enable(gradient_checkpointing_kwargs={"preserve_rng_state": False})
+        else:
```
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/cloud/finetune.py:115]: Fixed in latest

File: QEfficient/finetune/utils/train_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py]: Removed in latest.

**Review by quic-swatia — APPROVED** [review]

**Review by quic-mamta — APPROVED** [review]
